### ISSUE TITLE
Unify 3mfverify.exe with xml-linting in lib3MF-repo

### ISSUE THREAD

**Issue by martinweismann** [3MFConsortium/3mf-samples#11]
https://github.com/3MFConsortium/lib3mf/tree/XML_Schema_Validation/Tools/Schema 
contains scripts to perform xml-schema validation for an individual 3dmodel.model against the xsd-files from the different published extensions. 

3mfverify.exe appears to verify much more (OPC structure,…)  but does not include the latest extensions.

@pencerw : Let's work on unifying the functionality from these tools.
Do you know any internals about the 3mfverify-tool? It might suffice to feed the correct schemata into it.

**Comment by pencerw** [comment]
@martinweismann unfortunately I came into this repo only in January, and know nothing about the source of these files: https://github.com/3MFConsortium/3mf-samples/tree/master/3mf-Verify

It appears that @pabloprado added them to the repository - perhaps he or @alexanderoster knows?

**Comment by thatrani** [comment]
@sawhite500 , is there a replacement for 3mfverify.exe that HP created?

**Comment by ghost** [comment]
> @sawhite500 , is there a replacement for 3mfverify.exe that HP created?

I just checked - there hasn't been any activity on that tool for well over a year.  Is there a need / request to add capability from the POV of the consortium?

**Comment by KrisIverson** [comment]
I do not clearly recall, but I thought MS supplied the 3mfverify tool? I remember adding code to URL link to 3MF documentation when errors were detected. I do not know if anyone is using this tool. We had aspirations to move it to an online service or expand https://3mf.azurewebsite.net with this functionality but did not make progress towards that.

-K

**Comment by ghost** [comment]
Good point Kris - I don’t remember exactly how it all went down.  We did, I believe, share the code from HP’s verification tool with MSFT.  It may be that it was never used (it’s written in Ruby).

Thanks,

Scott

**Comment by KrisIverson** [comment]
What kind of tool do we want for the future?

**Comment by martinweismann** [comment]
I see two main users of such a tool:

- As an implementer of the 3MF standard I would like a tool that, given a 3MF file from someone else, tells me, whether it is valid (do i have to accept this file?) and whether files created by my tool are valid (do others have to accept this file?)
- As a user working with 3MF files, I want to be able to point others (producers, consumers) to an authoritative tool that can judge whether a 3mf-file is valid. This way, interoperation between tools can be put on very solid grounds without deep understanding of the 3MF specifications. (Obviously, this does not replace detailed spec knowledge if errors or interop-problems arise)

What kind of tool should this be:
- 3mfverify was not to bad at that. As far as I remember it verifies the OPC package structure and verifies, whether the 3dmodel and the relationship-files are a valid instances of their respective schemata. However, it is limited to the core spec, and it did not respect any "soft parts" of the core specification, i.e. wording of the specification that can not be backed into an XSD, e.g. one MUST not reference resource IDs of objects defined later in a file.
- Arguably, one can say, that these capabilities fall into the scope of a reference implementation itself. However, tools reading a file type are usually held to be forgiving about errors in files. E.g. XML linting and verification of schema conformance should not part of the parsing a 3MF file, since this is 
a) a deteriorates performance
b) might lead to rejection of files that could easily be understood by others although they violate the schema (e.g. unknown attributes of an element within the namespace of the core-spec).

**Comment by thatrani** [comment]
I would also add to the wishlist that it is a web service.
